# This notebook includes the short Part 1 of the assignment,
# an even shorter Part 2 and a very long Part 3

# Part 1

*I spent an hour to figure out web scraping, but I did not understand how to get the specific page structure of 'M' only.*

*To fasten the read process, I have copied the table from Wikipedia into Excel and saved it in csv format.*

In [1]:
# import pandas and dataframe
import pandas as pd

df = pd.read_csv('Post_Codes_Toronto.csv', delimiter = ';')
# Use Postal Code as index to ease concatination with geospatial data
df.set_index('Postal Code', inplace = True)

In [2]:
# quick check dataframe
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 180 entries, M1A to M9Z
Data columns (total 2 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   Borough        180 non-null    object
 1   Neighbourhood  180 non-null    object
dtypes: object(2)
memory usage: 4.2+ KB


In [3]:
# check Boroughs for 'Not assigned'
df['Borough'].value_counts()

Not assigned        77
North York          24
Downtown Toronto    19
Scarborough         17
Etobicoke           12
Central Toronto      9
West Toronto         6
East York            5
York                 5
East Toronto         5
Mississauga          1
Name: Borough, dtype: int64

77 Boroughs not assigned.

In [4]:
# drop Boroughs 'Not assigned'
df = df[df.Borough != 'Not assigned']

In [5]:
# check Neighbourhood for 'Not assigned'
df['Neighbourhood'].value_counts()

Downsview                                                                                                                                 4
Don Mills                                                                                                                                 2
Studio District                                                                                                                           1
Steeles West, L'Amoreaux West                                                                                                             1
Guildwood, Morningside, West Hill                                                                                                         1
                                                                                                                                         ..
East Toronto, Broadview North (Old East York)                                                                                             1
Moore Park, Summerhi

In [6]:
# check more specific Neighbourhood for 'Not assigned'
df[df.Neighbourhood == 'Not assigned']

Empty DataFrame
Columns: [Borough, Neighbourhood]
Index: []

No Neighbourhood not assigned.

In [7]:
# shape of dataframe
df.shape

(103, 2)

Shape: 103 rows in 2 columns (column 'Postal Code' as index is not counted)

# Part 2

-> geocoder did not run with the provided code skeleton, so another csv

In [8]:
df_geo = pd.read_csv('Geospatial_Coordinates.csv')
# Use Postal Code as index to ease concatination with main data
df_geo.set_index('Postal Code', inplace = True)

In [9]:
# Concatinate main data and geospatial data according to index
df_all = pd.concat([df, df_geo], axis = 1)
df_all.head()

Borough                           Neighbourhood   Latitude  \
Postal Code                                                                   
M1B          Scarborough                          Malvern, Rouge  43.806686   
M1C          Scarborough  Rouge Hill, Port Union, Highland Creek  43.784535   
M1E          Scarborough       Guildwood, Morningside, West Hill  43.763573   
M1G          Scarborough                                  Woburn  43.770992   
M1H          Scarborough                               Cedarbrae  43.773136   

             Longitude  
Postal Code             
M1B         -79.194353  
M1C         -79.160497  
M1E         -79.188711  
M1G         -79.216917  
M1H         -79.239476

Needed table with Postal Code, Borough, Neighbourhood, Latitude and Longitude

# Part 3

In [11]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

%pip install geopy
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

%pip install folium
import folium # map rendering library

Note: you may need to restart the kernel to use updated packages.
Note: you may need to restart the kernel to use updated packages.


In [16]:
# get geo data for Toronto map
address = 'Toronto'
geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude

# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighbourhood in zip(df_all['Latitude'], df_all['Longitude'], df_all['Borough'], df_all['Neighbourhood']):
    label = '{}, {}'.format(neighbourhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

In [17]:
CLIENT_ID = 'LR42PPMSIQLLUL4CQWG2FGYTRRHRTTKAWJPNCEEWOTXCX1Z3' # your Foursquare ID
CLIENT_SECRET = '2NJLDIH30EGQJK0WHDO3SCLEE5X154RZO4ZUDFUHOYEGYQTY' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version
LIMIT = 100 # A default Foursquare API limit value

In [18]:
#get venues in Toronto's neighbourhoods
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [20]:
toronto_venues = getNearbyVenues(names=df_all['Neighbourhood'],
                                   latitudes=df_all['Latitude'],
                                   longitudes=df_all['Longitude']
                                  )

Malvern, Rouge
Rouge Hill, Port Union, Highland Creek
Guildwood, Morningside, West Hill
Woburn
Cedarbrae
Scarborough Village
Kennedy Park, Ionview, East Birchmount Park
Golden Mile, Clairlea, Oakridge
Cliffside, Cliffcrest, Scarborough Village West
Birch Cliff, Cliffside West
Dorset Park, Wexford Heights, Scarborough Town Centre
Wexford, Maryvale
Agincourt
Clarks Corners, Tam O'Shanter, Sullivan
Milliken, Agincourt North, Steeles East, L'Amoreaux East
Steeles West, L'Amoreaux West
Upper Rouge
Hillcrest Village
Fairview, Henry Farm, Oriole
Bayview Village
York Mills, Silver Hills
Willowdale, Newtonbrook
Willowdale, Willowdale East
York Mills West
Willowdale, Willowdale West
Parkwoods
Don Mills
Don Mills
Bathurst Manor, Wilson Heights, Downsview North
Northwood Park, York University
Downsview
Downsview
Downsview
Downsview
Victoria Village
Parkview Hill, Woodbine Gardens
Woodbine Heights
The Beaches
Leaside
Thorncliffe Park
East Toronto, Broadview North (Old East York)
The Danforth West, 

In [21]:
print(toronto_venues.shape)
toronto_venues.head()

(2117, 7)


Neighbourhood  Neighbourhood Latitude  \
0                          Malvern, Rouge               43.806686   
1                          Malvern, Rouge               43.806686   
2  Rouge Hill, Port Union, Highland Creek               43.784535   
3       Guildwood, Morningside, West Hill               43.763573   
4       Guildwood, Morningside, West Hill               43.763573   

   Neighbourhood Longitude                  Venue  Venue Latitude  \
0               -79.194353                Wendy’s       43.807448   
1               -79.194353  Interprovincial Group       43.805630   
2               -79.160497  Royal Canadian Legion       43.782533   
3               -79.188711         RBC Royal Bank       43.766790   
4               -79.188711      G & G Electronics       43.765309   

   Venue Longitude        Venue Category  
0       -79.199056  Fast Food Restaurant  
1       -79.200378            Print Shop  
2       -79.163085                   Bar  
3       -79.191151                  Bank  
4       -79.191537     Electronics Store

2117 venues collected in all neighbourhoods together

In [23]:
toronto_venues.groupby('Neighbourhood').count()

Neighbourhood Latitude  \
Neighbourhood                                                                
Agincourt                                                                5   
Alderwood, Long Branch                                                   7   
Bathurst Manor, Wilson Heights, Downsview North                         21   
Bayview Village                                                          4   
Bedford Park, Lawrence Manor East                                       25   
Berczy Park                                                             55   
Birch Cliff, Cliffside West                                              4   
Brockton, Parkdale Village, Exhibition Place                            24   
Business reply mail Processing Centre, South Ce...                      16   
CN Tower, King and Spadina, Railway Lands, Harb...                      17   
Caledonia-Fairbanks                                                      4   
Canada Post Gateway Processing Centre                                   15   
Cedarbrae                                                                9   
Central Bay Street                                                      62   
Christie                                                                16   
Church and Wellesley                                                    79   
Clarks Corners, Tam O'Shanter, Sullivan                                 14   
Cliffside, Cliffcrest, Scarborough Village West                          2   
Commerce Court, Victoria Hotel                                         100   
Davisville                                                              35   
Davisville North                                                         8   
Del Ray, Mount Dennis, Keelsdale and Silverthorn                         4   
Don Mills                                                               27   
Dorset Park, Wexford Heights, Scarborough Town ...                       7   
Downsview                                                               14   
Dufferin, Dovercourt Village                                            14   
East Toronto, Broadview North (Old East York)                            3   
Eringate, Bloordale Gardens, Old Burnhamthorpe,...                       8   
Fairview, Henry Farm, Oriole                                            60   
First Canadian Place, Underground city                                 100   
Forest Hill North & West, Forest Hill Road Park                          4   
Garden District, Ryerson                                               100   
Glencairn                                                                5   
Golden Mile, Clairlea, Oakridge                                         10   
Guildwood, Morningside, West Hill                                        8   
Harbourfront East, Union Station, Toronto Islands                      100   
High Park, The Junction South                                           25   
Hillcrest Village                                                        5   
Humber Summit                                                            1   
Humberlea, Emery                                                         1   
Humewood-Cedarvale                                                       4   
India Bazaar, The Beaches West                                          18   
Kennedy Park, Ionview, East Birchmount Park                              5   
Kensington Market, Chinatown, Grange Park                               66   
Kingsview Village, St. Phillips, Martin Grove G...                       4   
Lawrence Manor, Lawrence Heights                                        14   
Lawrence Park                                                            4   
Leaside                                                                 33   
Little Portugal, Trinity                                                42   
Malvern, Rouge                                                           2

In [25]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 276 uniques categories.


In [26]:
# one hot encoding to categorize venues
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighbourhood'] = toronto_venues['Neighbourhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Neighbourhood  Accessories Store  Adult Boutique  \
0                          Malvern, Rouge                  0               0   
1                          Malvern, Rouge                  0               0   
2  Rouge Hill, Port Union, Highland Creek                  0               0   
3       Guildwood, Morningside, West Hill                  0               0   
4       Guildwood, Morningside, West Hill                  0               0   

   Airport  Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0        0                   0             0               0                0   
1        0                   0             0               0                0   
2        0                   0             0               0                0   
3        0                   0             0               0                0   
4        0                   0             0               0                0   

   Airport Terminal  American Restaurant  Antique Shop  Aquarium  Art Gallery  \
0                 0                    0             0         0            0   
1                 0                    0             0         0            0   
2                 0                    0             0         0            0   
3                 0                    0             0         0            0   
4                 0                    0             0         0            0   

   Art Museum  Arts & Crafts Store  Asian Restaurant  Athletics & Sports  \
0           0                    0                 0                   0   
1           0                    0                 0                   0   
2           0                    0                 0                   0   
3           0                    0                 0                   0   
4           0                    0                 0                   0   

   Auto Garage  Auto Workshop  BBQ Joint  Baby Store  Bagel Shop  Bakery  \
0            0              0          0           0           0       0   
1            0              0          0           0           0       0   
2            0              0          0           0           0       0   
3            0              0          0           0           0       0   
4            0              0          0           0           0       0   

   Bank  Bar  Baseball Field  Baseball Stadium  Basketball Court  \
0     0    0               0                 0                 0   
1     0    0               0                 0                 0   
2     0    1               0                 0                 0   
3     1    0               0                 0                 0   
4     0    0               0                 0                 0   

   Basketball Stadium  Beach  Bed & Breakfast  Beer Bar  Beer Store  \
0                   0      0                0         0           0   
1                   0      0                0         0           0   
2                   0      0                0         0           0   
3                   0      0                0         0           0   
4                   0      0                0         0           0   

   Belgian Restaurant  Bike Shop  Bistro  Board Shop  Boat or Ferry  \
0                   0          0       0           0              0   
1                   0          0       0           0              0   
2                   0          0       0           0              0   
3                   0          0       0           0              0   
4                   0          0       0           0              0   

   Bookstore  Boutique  Brazilian Restaurant  Breakfast Spot  Brewery  \
0          0         0                     0               0        0   
1          0         0                     0               0        0   
2          0         0                     0               0        0   
3          0         0                     0               0        0   
4          0         0         

In [28]:
toronto_grouped = toronto_onehot.groupby('Neighbourhood').mean().reset_index()
toronto_grouped

Neighbourhood  Accessories Store  \
0                                           Agincourt           0.000000   
1                              Alderwood, Long Branch           0.000000   
2     Bathurst Manor, Wilson Heights, Downsview North           0.000000   
3                                     Bayview Village           0.000000   
4                   Bedford Park, Lawrence Manor East           0.000000   
5                                         Berczy Park           0.000000   
6                         Birch Cliff, Cliffside West           0.000000   
7        Brockton, Parkdale Village, Exhibition Place           0.000000   
8   Business reply mail Processing Centre, South C...           0.000000   
9   CN Tower, King and Spadina, Railway Lands, Har...           0.000000   
10                                Caledonia-Fairbanks           0.000000   
11              Canada Post Gateway Processing Centre           0.000000   
12                                          Cedarbrae           0.000000   
13                                 Central Bay Street           0.000000   
14                                           Christie           0.000000   
15                               Church and Wellesley           0.000000   
16            Clarks Corners, Tam O'Shanter, Sullivan           0.000000   
17    Cliffside, Cliffcrest, Scarborough Village West           0.000000   
18                     Commerce Court, Victoria Hotel           0.000000   
19                                         Davisville           0.000000   
20                                   Davisville North           0.000000   
21   Del Ray, Mount Dennis, Keelsdale and Silverthorn           0.000000   
22                                          Don Mills           0.000000   
23  Dorset Park, Wexford Heights, Scarborough Town...           0.000000   
24                                          Downsview           0.000000   
25                       Dufferin, Dovercourt Village           0.000000   
26      East Toronto, Broadview North (Old East York)           0.000000   
27  Eringate, Bloordale Gardens, Old Burnhamthorpe...           0.000000   
28                       Fairview, Henry Farm, Oriole           0.016667   
29             First Canadian Place, Underground city           0.000000   
30    Forest Hill North & West, Forest Hill Road Park           0.000000   
31                           Garden District, Ryerson           0.000000   
32                                          Glencairn           0.000000   
33                    Golden Mile, Clairlea, Oakridge           0.000000   
34                  Guildwood, Morningside, West Hill           0.000000   
35  Harbourfront East, Union Station, Toronto Islands           0.000000   
36                      High Park, The Junction South           0.000000   
37                                  Hillcrest Village           0.000000   
38                                      Humber Summit           0.000000   
39                                   Humberlea, Emery           0.000000   
40                                 Humewood-Cedarvale           0.000000   
41                     India Bazaar, The Beaches West           0.000000   
42        Kennedy Park, Ionview, East Birchmount Park           0.000000   
43          Kensington Market, Chinatown, Grange Park           0.000000   
44  Kingsview Village, St. Phillips, Martin Grove ...           0.000000   
45                   Lawrence Manor, Lawrence Heights           0.071429   
46                                      Lawrence Park           0.000000   
47                                            Leaside           0.000000   
48                           Little Portugal, Trinity           0.000000   
49                                     Malvern, Rouge           0.000000   
50  Milliken, Agincourt North, Steeles East, L'Amo...           0.000000   
51  Mimico NW, The Queensway West, South of Bloor,...           0.000000   
52          

In [29]:
num_top_venues = 5

for hood in toronto_grouped['Neighbourhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighbourhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Agincourt----
                       venue  freq
0  Latin American Restaurant   0.2
1             Breakfast Spot   0.2
2                     Lounge   0.2
3             Clothing Store   0.2
4               Skating Rink   0.2


----Alderwood, Long Branch----
            venue  freq
0     Pizza Place  0.29
1             Gym  0.14
2    Skating Rink  0.14
3  Sandwich Place  0.14
4     Coffee Shop  0.14


----Bathurst Manor, Wilson Heights, Downsview North----
               venue  freq
0               Bank  0.10
1        Coffee Shop  0.10
2           Pharmacy  0.05
3         Restaurant  0.05
4  Mobile Phone Shop  0.05


----Bayview Village----
                 venue  freq
0  Japanese Restaurant  0.25
1                 Bank  0.25
2   Chinese Restaurant  0.25
3                 Café  0.25
4    Accessories Store  0.00


----Bedford Park, Lawrence Manor East----
                venue  freq
0         Coffee Shop  0.08
1      Sandwich Place  0.08
2     Thai Restaurant  0.08
3  Italian Restaura

             venue  freq
0      Coffee Shop  0.12
1             Café  0.05
2       Restaurant  0.04
3  Thai Restaurant  0.03
4           Bakery  0.03


----Rosedale----
               venue  freq
0               Park  0.50
1         Playground  0.25
2              Trail  0.25
3  Mobile Phone Shop  0.00
4              Motel  0.00


----Roselawn----
                      venue  freq
0      Fast Food Restaurant   0.5
1                    Garden   0.5
2            Medical Center   0.0
3  Mediterranean Restaurant   0.0
4               Men's Store   0.0


----Rouge Hill, Port Union, Highland Creek----
                        venue  freq
0                         Bar   1.0
1           Accessories Store   0.0
2  Modern European Restaurant   0.0
3               Movie Theater   0.0
4                       Motel   0.0


----Runnymede, Swansea----
              venue  freq
0              Café  0.09
1       Coffee Shop  0.09
2  Sushi Restaurant  0.06
3               Pub  0.06
4       Pizza Place  0

In [30]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [108]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighbourhood'] = toronto_grouped['Neighbourhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighbourhood 1st Most Common Venue  \
0                                        Agincourt          Skating Rink   
1                           Alderwood, Long Branch           Pizza Place   
2  Bathurst Manor, Wilson Heights, Downsview North                  Bank   
3                                  Bayview Village    Chinese Restaurant   
4                Bedford Park, Lawrence Manor East    Italian Restaurant   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                Lounge        Breakfast Spot        Clothing Store   
1          Skating Rink           Coffee Shop                   Pub   
2           Coffee Shop           Pizza Place         Deli / Bodega   
3                  Café                  Bank   Japanese Restaurant   
4           Coffee Shop       Thai Restaurant        Sandwich Place   

       5th Most Common Venue 6th Most Common Venue   7th Most Common Venue  \
0  Latin American Restaurant            Donut Shop                   Diner   
1             Sandwich Place                   Gym  Gluten-free Restaurant   
2           Sushi Restaurant          Intersection       Mobile Phone Shop   
3                Yoga Studio    Dim Sum Restaurant                   Diner   
4                   Pharmacy               Butcher        Sushi Restaurant   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0        Discount Store   Distribution Center                Dog Run  
1          Cupcake Shop            Donut Shop       Doner Restaurant  
2           Supermarket        Ice Cream Shop    Fried Chicken Joint  
3        Discount Store   Distribution Center                Dog Run  
4                  Café     Indian Restaurant                    Pub

After several trials I have decided for only three clusters: A larger number of clusters leads to clusters with one single neigbourhood inside.

In [109]:
# set number of clusters
kclusters = 3

toronto_grouped_clustering = toronto_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=42).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10]

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0])

In [110]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = df_all

# merge manhattan_grouped with manhattan_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')

toronto_merged.head() # check the last columns!

Borough                           Neighbourhood   Latitude  \
Postal Code                                                                   
M1B          Scarborough                          Malvern, Rouge  43.806686   
M1C          Scarborough  Rouge Hill, Port Union, Highland Creek  43.784535   
M1E          Scarborough       Guildwood, Morningside, West Hill  43.763573   
M1G          Scarborough                                  Woburn  43.770992   
M1H          Scarborough                               Cedarbrae  43.773136   

             Longitude  Cluster Labels 1st Most Common Venue  \
Postal Code                                                    
M1B         -79.194353             0.0  Fast Food Restaurant   
M1C         -79.160497             0.0                   Bar   
M1E         -79.188711             0.0          Intersection   
M1G         -79.216917             0.0           Coffee Shop   
M1H         -79.239476             0.0                Bakery   

             2nd Most Common Venue 3rd Most Common Venue  \
Postal Code                                                
M1B                     Print Shop           Yoga Studio   
M1C                    Yoga Studio            Donut Shop   
M1E            Rental Car Location        Medical Center   
M1G          Korean BBQ Restaurant           Yoga Studio   
M1H                         Lounge      Hakka Restaurant   

            4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
Postal Code                                                                     
M1B              Doner Restaurant          Dessert Shop    Dim Sum Restaurant   
M1C                         Diner        Discount Store   Distribution Center   
M1E                Breakfast Spot    Mexican Restaurant                  Bank   
M1G              Doner Restaurant    Dim Sum Restaurant                 Diner   
M1H                          Bank   Fried Chicken Joint       Thai Restaurant   

            7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
Postal Code                                                                     
M1B                         Diner        Discount Store   Distribution Center   
M1C                       Dog Run      Doner Restaurant             Drugstore   
M1E                    Restaurant     Electronics Store     Convenience Store   
M1G                Discount Store   Distribution Center               Dog Run   
M1H            Athletics & Sports           Gas Station  Caribbean Restaurant   

            10th Most Common Venue  
Postal Code                         
M1B                        Dog Run  
M1C                   Dessert Shop  
M1E                     Comic Shop  
M1G                     Donut Shop  
M1H            Distribution Center

Drop unclustered rows and format cluster labels as int to enable map creation. (error message on floats)

In [111]:
toronto_merged = toronto_merged.dropna(subset=['Cluster Labels'], axis=0)

In [112]:
toronto_merged[['Cluster Labels']] = toronto_merged[['Cluster Labels']].astype('int')

In [113]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighbourhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

In [114]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Neighbourhood  \
Postal Code                                                      
M1B                                             Malvern, Rouge   
M1C                     Rouge Hill, Port Union, Highland Creek   
M1E                          Guildwood, Morningside, West Hill   
M1G                                                     Woburn   
M1H                                                  Cedarbrae   
M1K                Kennedy Park, Ionview, East Birchmount Park   
M1L                            Golden Mile, Clairlea, Oakridge   
M1M            Cliffside, Cliffcrest, Scarborough Village West   
M1N                                Birch Cliff, Cliffside West   
M1P          Dorset Park, Wexford Heights, Scarborough Town...   
M1R                                          Wexford, Maryvale   
M1S                                                  Agincourt   
M1T                    Clarks Corners, Tam O'Shanter, Sullivan   
M1W                              Steeles West, L'Amoreaux West   
M2H                                          Hillcrest Village   
M2J                               Fairview, Henry Farm, Oriole   
M2K                                            Bayview Village   
M2L                                   York Mills, Silver Hills   
M2N                                Willowdale, Willowdale East   
M2R                                Willowdale, Willowdale West   
M3A                                                  Parkwoods   
M3B                                                  Don Mills   
M3C                                                  Don Mills   
M3H            Bathurst Manor, Wilson Heights, Downsview North   
M3J                            Northwood Park, York University   
M3K                                                  Downsview   
M3L                                                  Downsview   
M3M                                                  Downsview   
M3N                                                  Downsview   
M4A                                           Victoria Village   
M4B                            Parkview Hill, Woodbine Gardens   
M4C                                           Woodbine Heights   
M4E                                                The Beaches   
M4G                                                    Leaside   
M4H                                           Thorncliffe Park   
M4K                               The Danforth West, Riverdale   
M4L                             India Bazaar, The Beaches West   
M4M                                            Studio District   
M4N                                              Lawrence Park   
M4P                                           Davisville North   
M4R                          North Toronto West, Lawrence Park   
M4S                                                 Davisville   
M4T                                Moore Park, Summerhill East   
M4V          Summerhill West, Rathnelly, South Hill, Forest...   
M4X                                St. James Town, Cabbagetown   
M4Y                                       Church and Wellesley   
M5A                                  Regent Park, Harbourfront   
M5B                                   Garden District, Ryerson   
M5C                                             St. James Town   
M5E                                                Berczy Park   
M5G                                         Central Bay Street   
M5H                                   Richmond, Adelaide, King   
M5J          Harbourfront East, Union Station, Toronto Islands   
M5K                   Toronto Dominion Centre, Design Exchange   
M5L                             Commerce Court, Victoria Hotel   
M5M                          Bedford Park, Lawrence Manor East   
M5N                                                   Roselawn   
M5P            Forest Hill North & West, Forest Hill Road Park   
M5R                        The Annex, North Midtown, Yorkville   
M5S                    

-> Cluster 0 looks like "hot spots" for young people with Bars, Diners and all kind of restaurants and shopping areas.

In [115]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Neighbourhood  \
Postal Code                                                      
M1J                                        Scarborough Village   
M1V          Milliken, Agincourt North, Steeles East, L'Amo...   
M9L                                              Humber Summit   

            1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
Postal Code                                                                     
M1J                    Playground               Dog Run      Department Store   
M1V                    Playground                  Park      Asian Restaurant   
M9L                  Intersection           Yoga Studio         Deli / Bodega   

            4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
Postal Code                                                                     
M1J                  Dessert Shop    Dim Sum Restaurant                 Diner   
M1V                  Intersection  Colombian Restaurant       College Stadium   
M9L                  Dessert Shop    Dim Sum Restaurant                 Diner   

            7th Most Common Venue 8th Most Common Venue  \
Postal Code                                               
M1J                Discount Store   Distribution Center   
M1V          Ethiopian Restaurant     Electronics Store   
M9L                Discount Store   Distribution Center   

                   9th Most Common Venue 10th Most Common Venue  
Postal Code                                                      
M1J                     Doner Restaurant     Falafel Restaurant  
M1V          Eastern European Restaurant    Dumpling Restaurant  
M9L                              Dog Run       Doner Restaurant

-> Cluster 1 looks like "multi-cultural" family areas with playgrounds and different kind of restaurants from all over the world.

In [116]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Neighbourhood  \
Postal Code                                                  
M2M                                Willowdale, Newtonbrook   
M2P                                        York Mills West   
M4J          East Toronto, Broadview North (Old East York)   
M4W                                               Rosedale   
M6E                                    Caledonia-Fairbanks   
M8X          The Kingsway, Montgomery Road, Old Mill North   
M9N                                                 Weston   

            1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
Postal Code                                                                     
M2M                          Park           Yoga Studio      Doner Restaurant   
M2P             Convenience Store                  Park           Yoga Studio   
M4J                          Park     Convenience Store           Yoga Studio   
M4W                          Park            Playground                 Trail   
M6E                          Park         Women's Store                  Pool   
M8X                         River                  Park           Yoga Studio   
M9N                          Park     Convenience Store         Jewelry Store   

            4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
Postal Code                                                                     
M2M                  Dessert Shop    Dim Sum Restaurant                 Diner   
M2P                  Dessert Shop    Dim Sum Restaurant                 Diner   
M4J                  Dessert Shop    Dim Sum Restaurant                 Diner   
M4W           Distribution Center      Department Store          Dessert Shop   
M6E              Doner Restaurant          Dessert Shop    Dim Sum Restaurant   
M8X                       Dog Run          Dessert Shop    Dim Sum Restaurant   
M9N                    Donut Shop                 Diner        Discount Store   

            7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
Postal Code                                                                     
M2M                Discount Store   Distribution Center               Dog Run   
M2P                Discount Store   Distribution Center               Dog Run   
M4J                Discount Store   Distribution Center               Dog Run   
M4W            Dim Sum Restaurant                 Diner        Discount Store   
M6E                         Diner        Discount Store   Distribution Center   
M8X                         Diner        Discount Store   Distribution Center   
M9N           Distribution Center               Dog Run      Doner Restaurant   

            10th Most Common Venue  
Postal Code                         
M2M                     Donut Shop  
M2P               Doner Restaurant  
M4J               Doner Restaurant  
M4W                        Dog Run  
M6E                        Dog Run  
M8X               Doner Restaurant  
M9N                    Yoga Studio

-> Cluster 2 looks like "privileged" areas with parks, Yoga Studios, Diners and Dessert Shops.